File: src/test/java/com/flexpoker/table/command/aggregate/singlehand/threeplayer/ThreePlayerSeatPositionsAndButtonAndBlindsTest.java
package com.flexpoker.table.command.aggregate.singlehand.threeplayer;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import java.util.HashSet;
import java.util.UUID;
import java.util.stream.Collectors;

import org.junit.Test;

import com.flexpoker.table.command.aggregate.testhelpers.TableTestUtils;
import com.flexpoker.table.command.events.HandDealtEvent;
import com.flexpoker.table.command.events.TableCreatedEvent;

public class ThreePlayerSeatPositionsAndButtonAndBlindsTest {

    @Test
    public void test() {
        var tableId = UUID.randomUUID();
        var player1Id = UUID.randomUUID();
        var player2Id = UUID.randomUUID();
        var player3Id = UUID.randomUUID();

        var table = TableTestUtils.createBasicTableAndStartHand(tableId, player1Id, player2Id, player3Id);

        // check seat positions
        var seatPositionToPlayerIdMap = ((TableCreatedEvent) table.fetchNewEvents().get(0)).getSeatPositionToPlayerMap();
        seatPositionToPlayerIdMap = seatPositionToPlayerIdMap.entrySet().stream()
                .filter(x -> x.getValue() != null)
                .collect(Collectors.toMap(x -> x.getKey(), x -> x.getValue()));

        var player1MatchList = seatPositionToPlayerIdMap.values().stream()
                .filter(x -> x.equals(player1Id)).collect(Collectors.toList());
        var player2MatchList = seatPositionToPlayerIdMap.values().stream()
                .filter(x -> x.equals(player2Id)).collect(Collectors.toList());
        var player3MatchList = seatPositionToPlayerIdMap.values().stream()
                .filter(x -> x.equals(player3Id)).collect(Collectors.toList());

        // verify that each player id is only in one seat position and that
        // those are the only three filled-in positions
        assertEquals(1, player1MatchList.size());
        assertEquals(1, player2MatchList.size());
        assertEquals(1, player3MatchList.size());

        var numberOfOtherPlayerPositions = seatPositionToPlayerIdMap.values().stream()
                .filter(x -> !x.equals(player1Id)).filter(x -> !x.equals(player2Id))
                .filter(x -> !x.equals(player3Id)).distinct().count();
        assertEquals(0, numberOfOtherPlayerPositions);

        // check blinds
        var player1Position = seatPositionToPlayerIdMap.entrySet().stream()
                .filter(x -> x.getValue().equals(player1Id)).findAny().get().getKey()
                .intValue();
        var player2Position = seatPositionToPlayerIdMap.entrySet().stream()
                .filter(x -> x.getValue().equals(player2Id)).findAny().get().getKey()
                .intValue();
        var player3Position = seatPositionToPlayerIdMap.entrySet().stream()
                .filter(x -> x.getValue().equals(player3Id)).findAny().get().getKey()
                .intValue();

        var handDealtEvent = ((HandDealtEvent) table.fetchNewEvents().get(2));

        assertFalse(handDealtEvent.getButtonOnPosition() == handDealtEvent.getSmallBlindPosition());
        assertFalse(handDealtEvent.getSmallBlindPosition() == handDealtEvent.getBigBlindPosition());
        assertFalse(handDealtEvent.getButtonOnPosition() == handDealtEvent.getBigBlindPosition());

        var buttonAndBlindPositions = new HashSet<>();
        buttonAndBlindPositions.add(Integer.valueOf(handDealtEvent.getButtonOnPosition()));
        buttonAndBlindPositions.add(Integer.valueOf(handDealtEvent.getSmallBlindPosition()));
        buttonAndBlindPositions.add(Integer.valueOf(handDealtEvent.getBigBlindPosition()));

        assertTrue(buttonAndBlindPositions.contains(Integer.valueOf(player1Position)));
        assertTrue(buttonAndBlindPositions.contains(Integer.valueOf(player2Position)));
        assertTrue(buttonAndBlindPositions.contains(Integer.valueOf(player3Position)));
    }

}

File: src/test/kotlin/com/flexpoker/table/command/aggregate/singlehand/threeplayer/ThreePlayerSeatPositionsAndButtonAndBlindsTest.kt
package com.flexpoker.table.command.aggregate.singlehand.threeplayer

import com.flexpoker.table.command.aggregate.testhelpers.createBasicTableAndStartHand
import com.flexpoker.table.command.events.HandDealtEvent
import com.flexpoker.table.command.events.TableCreatedEvent
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import java.util.HashSet
import java.util.UUID

class ThreePlayerSeatPositionsAndButtonAndBlindsTest {

    @Test
    fun test() {
        val tableId = UUID.randomUUID()
        val player1Id = UUID.randomUUID()
        val player2Id = UUID.randomUUID()
        val player3Id = UUID.randomUUID()
        val events = createBasicTableAndStartHand(tableId, player1Id, player2Id, player3Id)

        // check seat positions
        val seatPositionToPlayerIdMap = (events[0] as TableCreatedEvent).seatPositionToPlayerMap.filter { it.value != null }
        val player1MatchList = seatPositionToPlayerIdMap.values.filter { it == player1Id }
        val player2MatchList = seatPositionToPlayerIdMap.values.filter { it == player2Id }
        val player3MatchList = seatPositionToPlayerIdMap.values.filter { it == player3Id }

        // verify that each player id is only in one seat position and that
        // those are the only three filled-in positions
        assertEquals(1, player1MatchList.size)
        assertEquals(1, player2MatchList.size)
        assertEquals(1, player3MatchList.size)
        val numberOfOtherPlayerPositions = seatPositionToPlayerIdMap.values
            .filter { it != player1Id }
            .filter { it != player2Id }
            .filter { it != player3Id }
            .distinct().count()
        assertEquals(0, numberOfOtherPlayerPositions)

        // check blinds
        val player1Position = seatPositionToPlayerIdMap.entries.first { it.value == player1Id }.key
        val player2Position = seatPositionToPlayerIdMap.entries.first { it.value == player2Id }.key
        val player3Position = seatPositionToPlayerIdMap.entries.first { it.value == player3Id }.key
        val (_, _, _, _, _, _, buttonOnPosition, smallBlindPosition, bigBlindPosition) = events[2] as HandDealtEvent
        assertFalse(buttonOnPosition == smallBlindPosition)
        assertFalse(smallBlindPosition == bigBlindPosition)
        assertFalse(buttonOnPosition == bigBlindPosition)
        val buttonAndBlindPositions = HashSet<Any>()
        buttonAndBlindPositions.add(Integer.valueOf(buttonOnPosition))
        buttonAndBlindPositions.add(Integer.valueOf(smallBlindPosition))
        buttonAndBlindPositions.add(Integer.valueOf(bigBlindPosition))
        assertTrue(buttonAndBlindPositions.contains(player1Position))
        assertTrue(buttonAndBlindPositions.contains(player2Position))
        assertTrue(buttonAndBlindPositions.contains(player3Position))
    }

}